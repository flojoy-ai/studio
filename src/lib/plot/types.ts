export interface Drawable {
  draw: () => void;
}
