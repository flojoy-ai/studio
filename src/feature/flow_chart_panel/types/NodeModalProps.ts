import { NodeResult } from "@src/feature/common/types/ResultsType";
import { Node } from "reactflow";
import { ElementsData } from "@/types";

export type NodeModalProps = {
  modalIsOpen: boolean;
  closeModal: () => void;
  nodeLabel: string;
  nodeType: string;
  nd: NodeResult | null;
  pythonString: string;
  nodeFilePath: string;
  selectedNode: Node<ElementsData> | null;
};
