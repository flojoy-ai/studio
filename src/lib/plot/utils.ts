export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
