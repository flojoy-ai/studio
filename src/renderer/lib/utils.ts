import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
