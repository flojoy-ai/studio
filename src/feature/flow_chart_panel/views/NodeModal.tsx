import { Light as SyntaxHighlighter } from "react-syntax-highlighter";
import python from "react-syntax-highlighter/dist/cjs/languages/hljs/python";
import json from "react-syntax-highlighter/dist/cjs/languages/hljs/json";
import { JSONTree } from "react-json-tree";
import { PlotlyComponent, makePlotlyData } from "flojoy/components";
import { Flex, Box, Modal, createStyles, Button } from "@mantine/core";
import { MantineTheme, useMantineTheme } from "@mantine/styles";
import { NodeModalProps } from "../types/NodeModalProps";
import useKeyboardShortcut from "@src/hooks/useKeyboardShortcut";
import { useFlojoySyntaxTheme } from "@src/assets/FlojoyTheme";
import { NODES_REPO } from "@src/data/constants";
import { useMemo } from "react";

export const NodeModalStyles = createStyles((theme) => ({
  content: {
    borderRadius: 17,
    height: "700px",
  },
  header: {
    padding: "65px 80px 30px 80px",
    borderRadius: 17,
    position: "unset",
  },
  title: {
    display: "block",
    position: "relative",
    fontSize: 32,
    fontWeight: 700,
    color: `${theme.colors.title[0]}`,
  },
  body: {
    padding: "10px 65px 45px",
  },
  close: {
    position: "absolute",
    iconSize: 50,
    svg: {
      width: 40,
      height: 40,
    },
    right: 40,
    top: 45,
  },
  buttonStyle1: {
    width: 180,
    fontSize: 14,
    borderRadius: 35,
    backgroundColor: `${theme.colors.accent1[0]}`,
    borderColor: `${theme.colors.accent1[0]}`,
    border: "1px solid",
    color: `${
      theme.colorScheme === "dark" ? theme.colors.dark[7] : theme.colors.gray[0]
    }`,
    "&:hover": {
      backgroundColor: `${
        theme.colorScheme === "dark"
          ? theme.colors.accent1[2]
          : theme.colors.accent1[1]
      }`,
      color: `${
        theme.colorScheme === "dark"
          ? theme.colors.dark[7]
          : theme.colors.gray[0]
      }`,
    },
  },
  buttonStyle2: {
    width: 180,
    fontSize: 14,
    borderRadius: 35,
    color: theme.colors.accent1[0],
    borderColor: `${theme.colors.accent1[0]}`,
    backgroundColor: `${
      theme.colorScheme === "dark"
        ? theme.colors.accent4[1]
        : theme.colors.accent1[2]
    }`,
    border: "1px solid",
    "&:hover": {
      backgroundColor: `${
        theme.colorScheme === "dark"
          ? theme.colors.accent1[2]
          : theme.colors.accent1[1]
      }`,
      color: `${
        theme.colorScheme === "dark"
          ? theme.colors.dark[7]
          : theme.colors.gray[0]
      }`,
    },
  },
}));

const themeJSONTree = (theme: MantineTheme) => {
  const darkJSONTree = {
    scheme: "flojoy",
    author: "Jack",
    base00: `${theme.colors.modal[0]}`,
    base01: `${theme.colors.accent3[0]}`,
    base02: `${theme.colors.accent3[0]}`,
    base03: `${theme.colors.accent2[0]}`,
    base04: `${theme.colors.accent1[1]}`,
    base05: `${theme.colors.accent3[0]}`,
    base06: `${theme.colors.accent2[0]}`,
    base07: `${theme.colors.accent2[0]}`,
    base08: `${theme.colors.accent3[0]}`,
    base09: `${theme.colors.accent1[0]}`,
    base0A: `${theme.colors.accent1[1]}`,
    base0B: `${theme.colors.accent3[0]}`,
    base0C: `${theme.colors.text[0]}`,
    base0D: `${theme.colors.title[0]}`,
    base0E: `${theme.colors.accent1[0]}`,
    base0F: `${theme.colors.accent1[0]}`,
  };

  const lightJSONTree = {
    scheme: "flojoy",
    author: "Jack",
    base00: `${theme.colors.modal[0]}`,
    base01: `${theme.colors.accent3[0]}`,
    base02: `${theme.colors.accent3[0]}`,
    base03: `${theme.colors.accent2[0]}`,
    base04: `${theme.colors.accent1[0]}`,
    base05: `${theme.colors.accent3[0]}`,
    base06: `${theme.colors.accent2[0]}`,
    base07: `${theme.colors.accent2[0]}`,
    base08: `${theme.colors.accent3[0]}`,
    base09: `${theme.colors.accent1[0]}`,
    base0A: `${theme.colors.accent1[1]}`,
    base0B: `${theme.colors.accent3[0]}`,
    base0C: `${theme.colors.text[0]}`,
    base0D: `${theme.colors.title[0]}`,
    base0E: `${theme.colors.accent1[0]}`,
    base0F: `${theme.colors.accent1[0]}`,
  };
  return { darkJSONTree, lightJSONTree };
};

// Import only the languages needed to reduce bundle size
SyntaxHighlighter.registerLanguage("python", python);
SyntaxHighlighter.registerLanguage("json", json);

const NodeModal = ({
  modalIsOpen,
  closeModal,
  nodeLabel,
  nodeType,
  nd,
  nodeFilePath,
  pythonString,
  selectedNode,
}: NodeModalProps) => {
  const theme = useMantineTheme();
  const { classes } = NodeModalStyles();
  const { darkFlojoy, lightFlojoy } = useFlojoySyntaxTheme();
  const { lightJSONTree, darkJSONTree } = themeJSONTree(theme);

  const colorScheme = theme.colorScheme;

  useKeyboardShortcut("ctrl", "e", closeModal);
  useKeyboardShortcut("meta", "e", closeModal);
  const LINK = `${NODES_REPO}/${nodeFilePath
    .replace("\\", "/")
    .replace("PYTHON/nodes/", "")}`;

  return (
    <Modal
      data-testid="node-modal"
      opened={modalIsOpen}
      onClose={closeModal}
      size={800}
      title={nodeLabel}
      classNames={{
        content: classes.content,
        header: classes.header,
        title: classes.title,
        close: classes.close,
        body: classes.body,
      }}
    >
      <Modal.CloseButton data-testid="node-modal-closebtn" />

      <Flex gap="xl">
        <Box>
          <Button
            size="md"
            classNames={{ root: classes.buttonStyle2 }}
            component="a"
            href={LINK}
            target="_blank"
          >
            VIEW ON GITHUB
          </Button>
        </Box>
        <Box>
          <Button size="md" classNames={{ root: classes.buttonStyle2 }}>
            VIEW EXAMPLES
          </Button>
        </Box>
      </Flex>
      {nodeLabel !== undefined && nodeType !== undefined && (
        <h3
          style={{
            fontSize: 19,
            marginBottom: 5,
            lineHeight: 1.5,
            marginTop: 34,
            color: `${theme.colors.title[0]}`,
          }}
        >
          Function Type:{" "}
          <code style={{ color: `${theme.colors.accent1[0]}` }}>
            {nodeType}
          </code>
        </h3>
      )}
      {!nd?.result ? (
        <h3
          style={{
            fontSize: 19,
            fontWeight: 400,
            marginTop: 5,
            marginBottom: 15,
            color: `${theme.colors.text[0]}`,
          }}
        >
          <code>{nodeLabel}</code> not run yet - Click <i>Run Script</i>.
        </h3>
      ) : (
        <div>
          {nd?.result && (
            <PlotlyComponent
              id={nd.id}
              data={makePlotlyData(
                nd.result.default_fig.data,
                theme.colorScheme
              )}
              layout={{
                ...nd.result.default_fig.layout,
                title: nd.result.default_fig.layout?.title ?? nodeLabel,
              }}
              useResizeHandler
              style={{
                height: 635,
                width: 630,
              }}
              theme={theme.colorScheme}
            />
          )}
        </div>
      )}
      <div style={{ display: "flex" }}>
        <h2 style={{ fontSize: 22, marginTop: 2, marginBottom: 0 }}>
          Python code
        </h2>
      </div>
      <SyntaxHighlighter
        language="python"
        style={colorScheme === "dark" ? darkFlojoy : lightFlojoy}
      >
        {pythonString}
      </SyntaxHighlighter>
      <h2 style={{ fontSize: 22, marginTop: 28, marginBottom: 0 }}>
        Node data
      </h2>
      <div>
        <JSONTree
          data={selectedNode}
          theme={{
            extend: colorScheme === "dark" ? darkJSONTree : lightJSONTree,
            tree: {
              marginTop: 15,
              padding: 20,
            },
          }}
          labelRenderer={([key]) => (
            <span style={{ paddingLeft: 8 }}>{key}</span>
          )}
        ></JSONTree>
      </div>
    </Modal>
  );
};

export default NodeModal;
