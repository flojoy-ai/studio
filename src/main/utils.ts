import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
