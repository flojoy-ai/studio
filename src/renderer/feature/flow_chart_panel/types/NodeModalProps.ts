import { NodeResult } from "@src/feature/common/types/ResultsType";
import { Node } from "reactflow";
import { ElementsData } from "@src/types";
