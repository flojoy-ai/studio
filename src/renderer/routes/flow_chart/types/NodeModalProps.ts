import { NodeResult } from "@/renderer/routes/common/types/ResultsType";
import { Node } from "reactflow";
import { ElementsData } from "@/renderer/types";
