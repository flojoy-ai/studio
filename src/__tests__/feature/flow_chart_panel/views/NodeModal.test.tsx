import NodeModal from "@src/feature/flow_chart_panel/views/NodeModal";
import { NodeModalProps } from "@src/feature/flow_chart_panel/types/NodeModalProps";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

const props: NodeModalProps = {
  modalIsOpen: false,
  afterOpenModal: jest.fn(),
  closeModal: jest.fn(),
  modalStyles: {},
  nodeLabel: "test",
  nodeType: "test",
  nd: {
    cmd: "test",
    id: "test-1",
    result: {
      data: {
        type: "ordered_pair",
      },
      default_fig: {
        data: [],
      },
    },
  },
  pythonString: "test-python-string",
  defaultLayout: undefined,
  clickedElement: {},
};

jest.mock("@src/feature/common/PlotlyComponent", () => {
  const mockChild = jest
    .fn()
    .mockReturnValue(<div data-testid="plotly-component"></div>);
  return { __esModule: true, default: mockChild };
});

jest.mock("react-modal", () => {
  const ReactModal = jest
    .fn()
    .mockReturnValue(<div data-testid="react-modal" />);
  return {
    default: ReactModal,
  };
});

describe("NodeModal", () => {
  it("checks the snapshot", () => {
    const { container } = renderWithTheme(<NodeModal {...props} />);
    expect(container).toMatchSnapshot();
  });
  it("checks if the react modal component is rendered", () => {
    const { container, getByTestId } = renderWithTheme(
      <NodeModal {...props} />
    );

    const component = getByTestId("react-modal");

    expect(component).toBeInTheDocument();
  });
});
