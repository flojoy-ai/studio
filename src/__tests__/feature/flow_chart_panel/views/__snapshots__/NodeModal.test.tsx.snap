// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeModal checks the snapshot 1`] = `
<div>
  <div
    data-testid="react-modal"
  />
</div>
`;
